crates/subspace-farmer-components/src/segment_reconstruction.rs

```diff
@@ -28,84 +26,73 @@ pub enum SegmentReconstructionError {
 
-    let mut piece_requests = piece_indexes
-        .iter()
-        .map(|piece_index| async {
-            let _permit = match semaphore.clone().acquire_owned().await {
+    // This is so we can move references into the future below
+    let semaphore = &semaphore;
+    let acquired_pieces_counter = &acquired_pieces_counter;
+
+    let pieces = (starting_piece_index..)
+        .take(PIECES_IN_SEGMENT as usize)
+        .map(|piece_index| async move {
+            let _permit = match semaphore.acquire().await {
                 Ok(permit) => permit,
                 Err(error) => {
                     warn!(
-                        piece_index=*piece_index,
+                        piece_index,
                         %error,
                         "Semaphore was closed, interrupting piece recover..."
                     );
-                    return (*piece_index, Ok(None));
+                    return None;
                 }
             };
 
-            if acquired_pieces_counter.load(Ordering::Relaxed) >= required_pieces_number as usize {
-                trace!(piece_index = *piece_index, "Skipped piece acquiring.");
+            if acquired_pieces_counter.load(Ordering::SeqCst) >= required_pieces_number as usize {
+                trace!(piece_index, "Skipped piece acquiring.");
 
-                return (*piece_index, Ok(None));
+                return None;
             }
 
             let piece = piece_getter
-                .get_piece(*piece_index, PieceGetterRetryPolicy::NoRetry)
+                .get_piece(piece_index, PieceGetterRetryPolicy::Limited(0))
                 .await;
 
-            if let Ok(piece) = &piece {
-                if piece.is_some() {
-                    acquired_pieces_counter.fetch_add(1, Ordering::Relaxed);
+            match piece {
+                Ok(piece) => {
+                    if piece.is_some() {
+                        acquired_pieces_counter.fetch_add(1, Ordering::SeqCst);
+                    }
+
+                    piece
+                }
+                Err(error) => {
+                    debug!(?error, piece_index, "Failed to get piece");
+                    None
                 }
             }
-
-            (*piece_index, piece)
         })
-        .collect::<FuturesUnordered<_>>();
-
-    let mut pieces = iter::repeat(None)
-        .take(PIECES_IN_SEGMENT as usize)
-        .collect::<Vec<Option<Piece>>>();
-
-    while let Some(piece_response) = piece_requests.next().await {
-        match piece_response.1 {
-            Ok(piece) => {
-                let piece_index = piece_response.0;
-                let position = (piece_index - starting_piece_index) as usize;
-                pieces[position] = piece;
-            }
-            Err(err) => {
-                debug!(?err, "Piece getter failed.");
-            }
-        }
-    }
+        .collect::<FuturesOrdered<_>>()
+        .collect::<Vec<Option<Piece>>>()
```

Comment:
Hah, this `collect` method is so overpowered! Didn't know it was possible.

Reply:
The bigger point here was that you cared about order, but used unordered future with manual collection into ordered vector afterwards.

It wasn't necessary, without changing too much you could still use ordered future in a loop, but from there jump to simply collecting into the vector was obvious, so even if you don't know if it is possible, it kind of stimulates you to look into the docs, the pattern seems so natural that there should be something for it and there was indeed.